File: leet_BST_173_binary_search_tree_iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class BSTIterator:

    def __init__(self, node: Optional[TreeNode]):
        self.i = 0
        self.arr = []
        def inorder_traversal(self, node: Optional[TreeNode]):
            if not node:
                return
            inorder_traversal(self, node.left)
            self.arr.append(node.val)
            inorder_traversal(self, node.right)
        inorder_traversal(self, node)


    def next(self) -> int:
        n = len(self.arr)
        if self.i < n:
            val = self.arr[self.i]
        else:
            val = self.arr[n - 1]
        self.i += 1
        return val


    def hasNext(self) -> bool:
        return self.i < len(self.arr)


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: leet_BST_173_binary_search_tree_iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {

private:    

    void inorder_traversal(TreeNode* node)
    {
        if (!node)
            return ;
        inorder_traversal(node->left);
        arr.push_back(node->val);
        inorder_traversal(node->right);
    }

public:

    vector<int> arr;
    int         i = 0;

    BSTIterator(TreeNode* root)
    {
        inorder_traversal(root);
    }
    
    int next()
    {
        int n = (int) arr.size();
        if (i < n)
            return arr[i++];
        return arr[n - 1];
    }
    
    bool hasNext()
    {
        return i < arr.size();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */
